cilium/CFP-33716-packet-tracing.md

```diff
@@ -0,0 +1,636 @@
+
+## Motivation
+There are currently limitations for gathering and correlating flow events:
+- NAT transformations and DSR can all change the 4-tuple, so that a single Hubble or tcpdump filter cannot be used to trace the packet along the entire path – even when using a system such as Hubble Relay.
+- Enabling verbose datapath debugging is not possible in production environments due to traffic volume which leads to dropping flows.
+
+## Goals
+- Enable dataplane to watch for IP option tagged packets with `ip_trace_id` and emit flow events for tagged packets.
+- `ip_trace_id` can be used without enabling debug-verbose in the datapath.
+- Support both IPv4 and IPv6.
+- Add hubble filter for trace ID.
+- Allow users to define which flow IDs they want to monitor for events.
+
+## Non-goals
+- Automatically apply IP option to packets (see Future Milestones).
+
+## User Stories
+- As a platform administrator managing multiple clusters, users are reporting connection timeouts reaching destinations in remote clusters. Platform admin generates traffic with trace ID IP option along the problematic path to identify all flow events associated with that packet at both the source and destination.
+
+![User story](./images/CFP-33716-user-story.png)
+
+- As a production cluster administrator, particular flows from a client are experiencing connection timeouts (e.g. [issue #32472](https://github.com/cilium/cilium/issues/32472)). Enabling datapath debugging for all flows leads to event loss and low SNR. To extract dataplane information to root cause the issue, generate traffic with trace ID IP option to emit datapath debug information specific to this flow.
+
+## Proposal
+
+### Overview
+Offer support for packets containing a user-defined IP option field. For tagged incoming flows, handle events for this flow specially; emit datapath debugging events for these packets, include the ID in the event message, and implement hubble filtering by the ID.
+
+### Section 1: Feature Enablement
+
+When this feature is enabled, the datapath forwards packets to a parsing program on ingress.
+
+```go
+EnableIPOptionTracing = "enable-ip-option-tracing"
+```
+
+```go
+if option.Config.EnableIPOptionTracing > 0 {
+    cDefinesMap["ENABLE_PACKET_IP_TRACING"] = 
+    strconv.Itoa(option.Config.EnableIPOptionTracing)
+}
+```
+
+#### 1.1 DaemonSet-Level Enablement
+Enable IP option tracing for all agents in the cluster by setting the option at the DaemonSet level:
+- `--enable-ip-option-tracing=<type>` where `<type>` corresponds to the IP option value ([IANA IP Parameters](https://www.iana.org/assignments/ip-parameters/ip-parameters.xhtml)).
+- Default value: `<type>=0` (Feature is disabled).
+- When set at the DaemonSet level, all agents will embed the trace ID in the specified IP option type.
+
+Example DaemonSet configuration:
+```yaml
+args:
+  - --enable-ip-option-tracing=136
+```
+
+#### 1.2 Specific Agent-Level Enablement
```

Comment:
nit: 1.1 and 1.2 seem identical, and it's generally not meaningful to say whether config is in the daemonset or the agent. Configuration is specified in either a ConfigMap or in a CiliumNodeConfig CRD. I don't see the need to specify this twice.

```suggestion
#### 1.1 Configuration in cilium-agent
```

Reply:
Fixed and pushed.